src/openfermion/transforms/_conversion.py

```diff
@@ -556,3 +557,331 @@ def get_boson_operator(operator, hbar=1.):
+    reference determinant.
+
+    Args:
+        fermion_op(FermionOperator): An instance of the FermionOperator class.
+            It should not contain terms which do not preserve particle number.
+            If spin_preserving is set to True it should also not contain terms
+            which do not preserve the Sz (it is assumed that the ordering of
+            the indices goes alpha, beta, alpha, beta, ...).
+        num_qubits(int): The total number of qubits / spin-orbitals in the
+            system.
+        num_electrons(int): The number of particles in the desired Hilbert
+            space.
+        spin_preserving(bool): Whether or not the constructed operator should
+            be defined in a space which has support only on states with the
+            same Sz value as the reference_determinant.
+        reference_determinant(list(bool)): A list, whose length is equal to
+            num_qubits, which specifies which orbitals should be occupied in
+            the reference state. If spin_preserving is set to True then the Sz
+            value of this reference state determines the Sz value of the
+            symmetry sector in which the generated operator acts. If a value
+            for excitation_level is provided then the excitations are generated
+            with respect to the reference state. In any case, the ordering of
+            the states in the matrix representation of the operator depends on
+            reference_determinant and the state corresponding to
+            reference_determinant is the vector [1.0, 0.0, 0.0 ... 0.0]. Can be
+            set to None in order to take the first num_electrons orbitals to be
+            the occupied orbitals.
+        excitation_level(int): The number of excitations from the reference
+            state which should be included in the generated operator's matrix
+            representation. Can be set to None to include all levels of
+            excitation.
+
+    Returns:
+        sparse_op(scipy.sparse.csc_matrix): A sparse matrix representation of
+            fermion_op in the basis set by the arguments.
+    """
+
+    # We use the Hartree-Fock determinant as a reference if none is provided.
+    if reference_determinant is None:
+        reference_determinant = numpy.array([i < num_electrons for i in
+                                             range(num_qubits)])
+    else:
+        reference_determinant = numpy.asarray(reference_determinant)
+
+    if excitation_level is None:
+        excitation_level = num_electrons
+
+    state_array = numpy.asarray(list(_iterate_basis_(
+        reference_determinant, excitation_level, spin_preserving)))
+    # Create a 1d array with each determinant encoded as an integer for sorting purposes.
+    int_state_array = state_array.dot(1 << numpy.arange(state_array.shape[1])[::-1])
+    sorting_indices = numpy.argsort(int_state_array)
+
+    space_size = state_array.shape[0]
+
+    fermion_op = normal_ordered(fermion_op)
```

Comment:
Is the normal-ordering necessary? If not we should avoid it because it could be expensive.

Reply:
It dramatically simplifies the logic of other pieces and is necessary for the current implementation. With normal ordered terms we can just look at the presence or absence of certain operators and immediately infer whether or not a specific determinant has a non-zero matrix element.

Reply:
I would love to take on speeding up normal ordering in OpenFermion as a pet project though. I agree that it could be quite painful even for an otherwise reasonable operator.